### PR TITLE
Skip mask creation when padding_mask is None under torch.compile

### LINKED ISSUE #48925: 4D mask created under torch.compile when padding_mask is None
### Summary
When running models with `torch.compile` and unpadded inputs (`attention_mask=None`), 4D masks are still materialized instead of passing `attn_mask=None` to SDPA. This causes noticeable SDPA slowdown due to unnecessary mask-add operations.

### Root cause
In `src/transformers/masking_utils.py`:
```python
if is_tracing(padding_mask):
    return False
```
Under torch.compile, is_tracing(None) evaluates to True, which prevents skipping mask creation even when padding_mask is None.

### Expected behavior

When attention_mask=None, create_bidirectional_mask (and create_causal_mask) should return None so SDPA runs without a dense mask.

### PR DESCRIPTION
**Pull request by dafahaha**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48928) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48928)

Fixes #48925

When padding_mask is None, skip mask creation under torch.compile. is_tracing(None) returns True, which prevented skipping mask creation even when no mask is needed.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Hmn, I think this breaks export and might introduce a graph break in compilation! The idea is okay, but we want to make sure there there aren't any data-dependent checks in a tracing/export path. Maybe do purely shape-based decisions in the tracing path?

**Comment by dafahaha** [comment]
Good catch. I'll rework it to use a shape-based check instead of the explicit None comparison, so it doesn't introduce a data-dependent branch in the tracing path. Pushing an update shortly.

**Comment by dafahaha** [comment]
Good point. I noticed #48940 uses the same approach, and msnliu verified it with fullgraph=True and torch.export.export — both paths compiled without graph breaks. Let me know if you'd still prefer a different approach, otherwise I can add the second instance (_can_skip_bidirectional_mask_xpu) and run the same tests.

**Comment by dafahaha** [comment]
Added the second instance (_can_skip_bidirectional_mask_xpu) as well. 

Verification:
- torch.compile(fullgraph=True): both None and Tensor padding_mask paths compile without graph breaks
- torch.export.export: succeeds for both None and Tensor cases
- The `padding_mask is not None` check prevents is_tracing(None) from returning True, which was the root cause

Both instances now consistent. Ready for review.

**Comment by dafahaha** [comment]
I notice #48940 (msnliu's PR) has the exact same CI failures on the same two shards (tests_non_model shard 3 and 5). Since the changes are identical, these failures are likely either pre-existing flaky tests or an expected side effect of this fix. Happy to dig into the failing tests if you think they need to be addressed, but just wanted to flag that it's consistent across both PRs.

**Comment by dafahaha** [comment]
Reworked the approach: instead of checking `padding_mask is not None` (which is data-dependent), we now use `is_torchdynamo_exporting()` directly. This only blocks skipping under export (where is_causal is hard-coded), not under compile. So under torch.compile, when padding_mask is None, we can now skip mask creation as intended.

**Comment by dafahaha** [comment]
Heads up on the CI status: the failures on this PR are from the self-hosted runner infra (`Executing the custom container implementation failed. Please contact your self hosted runner administrator.`), not the diff itself — they hit both code-quality jobs before any tests ran.

Also noting that #48940 covers the same two lines with the same `is_torchdynamo_exporting()` approach and its full CI passed (190,190 tests, 0 failures). Since the fixes are identical, happy to close this one in favor of #48940 to keep the review queue clean — just let me know if there's anything in this version you'd prefer kept.

**Comment by zucchini-nlp** [comment]
It was fixed just now on `main`, so this one should be closed i believe
